### DISCUSSION TITLE
Use errors in custom middleware

### DISCUSSION THREAD

**Discussion by christopherpross** [dotnet/yarp#1198] [Q&A] [answered]
Hey,
I have some Middlewares in yarp and I want to generate errors and then modify my flow or not proxy the request.
I read about the "IProxyErrorFeature" but it doesn't exist anymore.
Is there an other way to do that?
all the best,
Christopher

**Comment by alnikola** [comment]
Hi. `IProxyErrorFeature` was renamed to `ForwaderErrorFeature`. Please, find it [here](https://github.com/microsoft/reverse-proxy/blob/main/src/ReverseProxy/Forwarder/IForwarderErrorFeature.cs).

**Reply by Tratcher** [reply]
Broken link?

**Reply by alnikola** [reply]
Sorry. My mistake. Fixed.

**Comment by christopherpross** [comment, marked as answer]
That means:
I need to implement the interface [IForwarderErrorFeature](https://raw.githubusercontent.com/microsoft/reverse-proxy/main/src/ReverseProxy/Forwarder/IForwarderErrorFeature.cs) and can then set / get it in my middleware to look whether there was are an exception or not?
Does yarp generate this also this way? I know that HttpForwarder generate this, but I only use middleware.
My target is, to react on any exception / error that can happen, custom or yarp. On the best way.

**Reply by alnikola** [reply]
I'm not sure if I completely understood your question, but I'd like to clarify that `IForwarderErrorFeature` is specific to `ForwarderMiddleware` and `HttpForwader` types. It's not a generic error reporting mechanism.

**Reply by MihaZupan** [reply]
YARP will implement the interface, you just have to get it inside your middleware. `HttpForwarder` will set the `IForwarderErrorFeature` on the context before returning, so you can inspect the error even if you are using the middleware and not the forwarder directly.

```c#
endpoints.MapReverseProxy(proxyPipeline =>
{
    proxyPipeline.Use(CustomMiddleware);
});

public async Task CustomMiddleware(HttpContext context, Func<Task> next)
{
    try
    {
        await next();

        var error = context.Features.Get<IForwarderErrorFeature>();
        if (error is not null)
        {
            // We couldn't proxy the request for some reason
            // May include the underlying exception
            Console.WriteLine($"Error: {error.Error} ({error.Exception?.Message})");
        }
    }
    catch (Exception ex)
    {
        // Something even more unexpected happened
        // For example a transform threw an exeception
    }
}
```
